lib/Conversion/TorchToLinalg/Uncategorized.cpp

```diff
@@ -307,6 +308,90 @@ Value createDivModePayload(OpBuilder &b, Location loc,
   return quotient;
 }
 
+template <typename OpT>
+Value createRemainderPayload(OpBuilder &b, Location loc,
+                             const TypeConverter *converter,
+                             ValueRange payloadArgs, OpT op,
+                             ArrayRef<Value> operands) {
+  static_assert(std::is_same_v<OpT, AtenRemainderScalarOp> ||
+                    std::is_same_v<OpT, AtenRemainderTensorOp>,
+                "template type must be a tensor/scalar remainder");
+  typename OpT::Adaptor adaptor(operands);
+  Type dtype = cast<RankedTensorType>(converter->convertType(op.getType()))
+                   .getElementType();
+  Value lhs = convertScalarToDtype(b, loc, payloadArgs[0], dtype);
+  Value rhs = convertScalarToDtype(
+      b, loc,
+      std::is_same_v<OpT, AtenRemainderScalarOp> ? operands[1] : payloadArgs[1],
+      dtype);
+
+  // The remainder op we wish to create would look roughly like this:
+  // rem = a % b
+  // if rem != 0 AND (rem < 0 XOR b < 0) rem += b
+  // This is how python calucates remainders for floats and longs:
+  // https://github.com/python/cpython/blob/2afd1751dd9a35d4ec03b708e3e5cddd72c43f7e/Objects/floatobject.c#L645
+  // https://github.com/python/cpython/blob/2afd1751dd9a35d4ec03b708e3e5cddd72c43f7e/Objects/longobject.c#L3662
+  Value result;
+  if (isa<mlir::FloatType>(dtype)) {
+    Value remainder = b.create<arith::RemFOp>(loc, lhs, rhs);
+
+    Value zero = b.create<arith::ConstantOp>(loc, b.getZeroAttr(dtype));
+    Value remainderNotEqualToZero = b.create<arith::CmpFOp>(
+        loc, arith::CmpFPredicate::ONE, remainder, zero);
+    Value otherLessThanZero =
+        b.create<arith::CmpFOp>(loc, arith::CmpFPredicate::OLT, rhs, zero);
+    Value remainderLessThanZero = b.create<arith::CmpFOp>(
+        loc, arith::CmpFPredicate::OLT, remainder, zero);
+    Value xorCondition =
+        b.create<arith::XOrIOp>(loc, otherLessThanZero, remainderLessThanZero);
+    Value condition =
+        b.create<arith::AndIOp>(loc, remainderNotEqualToZero, xorCondition);
+    result = b.create<scf::IfOp>(
+                  loc, condition,
+                  // if branch
+                  [&](OpBuilder &b, Location loc) {
+                    Value fixedRemainder =
+                        b.create<arith::AddFOp>(loc, remainder, rhs);
+                    b.create<scf::YieldOp>(loc, ValueRange(fixedRemainder));
+                  },
+                  // else branch
+                  [&](OpBuilder &b, Location loc) {
+                    b.create<scf::YieldOp>(loc, ValueRange(remainder));
+                  })
+                 .getResult(0);
```

Comment:
If I understand correctly, the pytorch remainder output matches the sign from the denominator, and the arith remainder op matches the sign from the numerator. Would it make sense to swap the numerator sign before taking the arith remainder, instead of fixing the value after the fact? E.g., something like:
1. `right_pos = (rhs > 0)` (arith.cmpf)
2. `poslhs = abs(lhs)` (math.absf)
3. `neglhs = -poslhs` (arith.negf)
4. `new_lhs = right_pos ? poslhs : neglhs` (arith.select)
5. `rem = new_lhs % rhs` (arith.remf)

Reply:
Yes, I just tried to check, and I think it would work, I copied the approach from python's codebase. I will try to modify it

Reply:
It seems that I've made a mistake, and actually this approach wouldn't work, for example.
if lhs = -13, and rhs = 10, the expected remainder is 7.
But this approach would give:
```
right_pos = 1
poslhs = 13
neglhs = -13
new_lhs = poslhs = 13
rem = 13 % 10 -> 3
```
The issue is that although the remainder matches the sign of the denominator, it's not an exact inverse of what the remainder would be if the numerator was positive. So it seems that I'll have to keep the old approach, although I'll probably modify it to use arith.select since it seems cleaner.

Reply:
I literally have a PhD in math, so this is pretty embarrassing for me haha. 

Thanks for looking into the suggestion even though it was bad!